### PR TITLE
OH2-172 add apis to enhance opd manipulation with operations rows as it is done for oh 1 11

### PR DESCRIPTION
**Pull request by ArnaudFonzam**
See OH2-172.

Paired with https://github.com/informatici/openhospital-core/pull/884.

### CODE REVIEW

File: src/main/java/org/isf/opd/dto/OpdWithOperatioRowDTO.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Open Hospital (www.open-hospital.org)
+ * Copyright © 2006-2022 Informatici Senza Frontiere (info@informaticisenzafrontiere.org)
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/dto/OpdWithOperatioRowDTO.java:3]:
```suggestion
 * Copyright © 2006-2023 Informatici Senza Frontiere (info@informaticisenzafrontiere.org)
```

File: src/main/java/org/isf/opd/rest/OpdController.java [hunk 2]
```diff
@@ -101,6 +120,45 @@ ResponseEntity<OpdDTO> newOpd(@RequestBody OpdDTO opdDTO) throws OHServiceExcept
...
+		opdToInsert.setPatient(patient);
+		opdToInsert.setWard(wardManager.findWard("OPD"));
+		Opd isCreatedOpd = opdManager.newOpd(opdToInsert);
+		if (isCreatedOpd == null) {
+			throw new OHAPIException(new OHExceptionMessage(null, "Opd is not created!", OHSeverityLevel.ERROR));
+		}
+		OpdDTO opdDTO = mapper.map2DTO(isCreatedOpd);
+		opdWithOperatioRow.setOpdDTO(opdDTO);
+		List<OperationRowDTO> listOp = new ArrayList<OperationRowDTO>();
+		if(opdWithOperatioRowDTO.getOperationRows().size() > 0) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/rest/OpdController.java:151]:
```suggestion
		if (opdWithOperatioRowDTO.getOperationRows().size() > 0) {
```

File: src/main/java/org/isf/opd/rest/OpdController.java [hunk 3]
```diff
@@ -101,6 +120,45 @@ ResponseEntity<OpdDTO> newOpd(@RequestBody OpdDTO opdDTO) throws OHServiceExcept
...
+	ResponseEntity<OpdWithOperatioRowDTO> newOpdWithOperationRow(@RequestBody OpdWithOperatioRowDTO opdWithOperatioRowDTO) throws OHServiceException {
+		int code = opdWithOperatioRowDTO.getOpdDTO().getCode();
+		LOGGER.info("store Out patient {}", code);
+		OpdWithOperatioRowDTO opdWithOperatioRow = new OpdWithOperatioRowDTO();
+		Patient patient = patientManager.getPatientById(opdWithOperatioRowDTO.getOpdDTO().getPatientCode());
+		if (patient == null) {
+			throw new OHAPIException(new OHExceptionMessage(null, "Patient not found!", OHSeverityLevel.ERROR));
+		}
+		if (opdWithOperatioRowDTO.getOpdDTO().getNote() == " ") {
+			throw new OHAPIException(new OHExceptionMessage(null, "not field is mandatory!", OHSeverityLevel.ERROR));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/rest/OpdController.java:139]:
```suggestion
			throw new OHAPIException(new OHExceptionMessage(null, "Note field is mandatory!", OHSeverityLevel.ERROR));
```

File: src/main/java/org/isf/opd/rest/OpdController.java [hunk 4]
```diff
@@ -133,6 +191,58 @@ ResponseEntity<OpdDTO> updateOpd(@PathVariable("code") int code, @RequestBody Op
...
+			throw new OHAPIException(new OHExceptionMessage(null, "Opd is not updated!", OHSeverityLevel.ERROR));
+		}
+		OpdDTO opdDTO = mapper.map2DTO(updatedOpd);
+		opdWithOperatioRow.setOpdDTO(opdDTO);
+		List<OperationRowDTO> listOpeRow = new ArrayList<OperationRowDTO>();
+		if(opdWithOperatioRowDTO.getOperationRows().size() > 0 ) {
+			for (OperationRowDTO operationRow : opdWithOperatioRowDTO.getOperationRows()) {
+				operationRow.setOpd(opdDTO);
+				OperationRow updated = new OperationRow();
+				if(operationRow.getId() == 0) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/rest/OpdController.java:233]:
```suggestion
				if (operationRow.getId() == 0) {
```

File: src/main/java/org/isf/opd/rest/OpdController.java [hunk 5]
```diff
@@ -101,6 +120,45 @@ ResponseEntity<OpdDTO> newOpd(@RequestBody OpdDTO opdDTO) throws OHServiceExcept
...
+	 * @throws OHServiceException
+	 */
+	@PostMapping(value = "/opds/rows", produces = MediaType.APPLICATION_JSON_VALUE)
+	ResponseEntity<OpdWithOperatioRowDTO> newOpdWithOperationRow(@RequestBody OpdWithOperatioRowDTO opdWithOperatioRowDTO) throws OHServiceException {
+		int code = opdWithOperatioRowDTO.getOpdDTO().getCode();
+		LOGGER.info("store Out patient {}", code);
+		OpdWithOperatioRowDTO opdWithOperatioRow = new OpdWithOperatioRowDTO();
+		Patient patient = patientManager.getPatientById(opdWithOperatioRowDTO.getOpdDTO().getPatientCode());
+		if (patient == null) {
+			throw new OHAPIException(new OHExceptionMessage(null, "Patient not found!", OHSeverityLevel.ERROR));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/rest/OpdController.java:136]: I personally find adding `!` to the end of an error message unnecessary and unprofessional.

Comment (ArnaudFonzam) [reply src/main/java/org/isf/opd/rest/OpdController.java:136]: for that i think it is better to create the task in other to remove all ! to the end of and error message

File: src/main/java/org/isf/opd/rest/OpdController.java [hunk 6]
```diff
@@ -199,25 +308,40 @@ public ResponseEntity<List<OpdDTO>> getOpdByDates(
...
-			opdDTOs = opds.stream().map(opd -> {
-				return mapper.map2DTO(opd);
+			opdWithOperations = opds.stream().map(opd -> {
+				OpdWithOperatioRowDTO opRows = new OpdWithOperatioRowDTO();
+				opRows.setOpdDTO(mapper.map2DTO(opd));
+				List<OperationRow> listOp = new ArrayList<OperationRow>();
+				try {
+					listOp = operationRowManager.getOperationRowByOpd(opd);
+				} catch (OHServiceException e) {
+					// TODO Auto-generated catch block
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/rest/OpdController.java:329]: This should be at a minimum a logger message to tract what happened.   Just dumping a stack trace is never a good idea.

File: src/main/java/org/isf/opd/rest/OpdController.java [hunk 7]
```diff
@@ -199,25 +308,40 @@ public ResponseEntity<List<OpdDTO>> getOpdByDates(
...
+					e.printStackTrace();
+				}
+				if(!listOp.isEmpty()) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/rest/OpdController.java:332]:
```suggestion
				if (!listOp.isEmpty()) {
```

File: src/main/java/org/isf/opd/rest/OpdController.java (hunk 5 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/rest/OpdController.java:136]: sorry @ArnaudFonzam how many "!" we have in "api" code so far?

Comment (mwithi) [reply src/main/java/org/isf/opd/rest/OpdController.java:136]: Ok, the task could be this one https://openhospital.atlassian.net/browse/OH2-182 that will also refactor how the errors are managed in API, without tampering with the core anymore.

File: src/main/java/org/isf/opd/rest/OpdController.java (hunk 6 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/rest/OpdController.java:329]: you can remove the comment //TODO then

**Comment by mwithi** [comment]
Please @ArnaudFonzam don't mix different branches !
Can you please force back this branch to https://github.com/informatici/openhospital-api/pull/215/commits/323e47b40f0a5a6261dd0106c708688e7879f31c ?

image

File: .gitignore [hunk 8]
```diff
@@ -1,5 +1,9 @@
 HELP.md
 /target/
+/rsc/application.properties
+/rsc/log4j.properties
+/rsc/settings.properties
+/rsc/database.properties
```
Comment (mwithi) [inline .gitignore:6]: This can be written as `/rsc/*.properties`

**Comment by mwithi** [comment]
> Please @ArnaudFonzam don't mix different branches ! Can you please force back this branch to [323e47b](https://github.com/informatici/openhospital-api/commit/323e47b40f0a5a6261dd0106c708688e7879f31c) ?
> 
> image

After you forceback, you can just realign this branch with actual develop (that now includes https://github.com/informatici/openhospital-api/pull/214 ). Thanks!

**Comment by mwithi** [comment]
> > Please @ArnaudFonzam don't mix different branches ! Can you please force back this branch to [323e47b](https://github.com/informatici/openhospital-api/commit/323e47b40f0a5a6261dd0106c708688e7879f31c) ?
> > image
> 
> After you forceback, you can just realign this branch with actual develop (that now includes #214 ). Thanks!

Since OP-179 has been merged "as is" the merge in this branch is totally fine and don't generate any conflict. 👍

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!

**Pull request merged**